tests/test_urban_nature_access.py

```diff
@@ -0,0 +1,134 @@
+        source_population_raster_path = os.path.join(
+            self.workspace_dir, 'population.tif')
+        population_pixel_size = (90, -90)
+        population_array_shape = (10, 10)
+
+        array_of_100s = numpy.full(
+            population_array_shape, 100, dtype=numpy.uint32)
+        array_of_random_ints = numpy.array(
+            random.choices(range(0, 100), k=100),
+            dtype=numpy.uint32).reshape(population_array_shape)
+
+        for population_array in (
+                array_of_100s, array_of_random_ints):
+            population_srs = osr.SpatialReference()
+            population_srs.ImportFromEPSG(_DEFAULT_EPSG)
+            population_wkt = population_srs.ExportToWkt()
+            pygeoprocessing.numpy_array_to_raster(
+                base_array=population_array,
+                target_nodata=-1,
+                pixel_size=population_pixel_size,
+                origin=_DEFAULT_ORIGIN,
+                projection_wkt=population_wkt,
+                target_path=source_population_raster_path)
+
+            for target_pixel_size in (
+                    (30, -30),  # 1/3 the pixel size
+                    (4, -4),  # way smaller
+                    (100, -100)):  # bigger
+                target_population_raster_path = os.path.join(
+                    self.workspace_dir, 'resampled_population.tif')
+                urban_nature_access._resample_population_raster(
+                    source_population_raster_path,
+                    target_population_raster_path,
+                    target_pixel_size=target_pixel_size,
+                    target_bb=pygeoprocessing.get_raster_info(
+                        source_population_raster_path)['bounding_box'],
+                    target_projection_wkt=population_wkt,
+                    working_dir=os.path.join(self.workspace_dir, 'working'))
+
+                resampled_population_array = (
+                    pygeoprocessing.raster_to_numpy_array(
+                        target_population_raster_path))
+
+                # There should be no significant loss or gain of population due
+                # to warping, but the fact that this is aggregating across the
+                # whole raster (lots of pixels) means we need to lower the
+                # relative tolerance.
+                numpy.testing.assert_allclose(
+                    population_array.sum(), resampled_population_array.sum(),
+                    rtol=1e-3)
+
+    def test_model(self):
+        """UNA: Run through the model."""
+        from natcap.invest import urban_nature_access
+
+        args = {
+            'workspace_dir': os.path.join(self.workspace_dir, 'workspace'),
+            'results_suffix': 'suffix',
+            'population_raster_path': os.path.join(
+                self.workspace_dir, 'population.tif'),
+            'lulc_raster_path': os.path.join(self.workspace_dir, 'lulc.tif'),
+        }
+
+        random.seed(-1)  # for our random number generation
```

Comment:
I'm not sure how we'll do the asserting yet, but I was wondering if we could rely on `random.seed` across different types of hardware / OS's?

Reply:
Ooo that's a fun question.  I didn't know the answer to it and didn't find anything while googling so I decided to test it out with everything Github Actions allows us to do while I would have loved to have tested every version of python that GHA provides, that wasn't practical (they only allow up to 256 tasks) so I had to settle for a subset, eventually using the latest bugfix (or alpha) releases for each major python version.  [Here are the results](https://github.com/phargogh/experiment-merseinne-twister-reliability/actions/runs/1469434143) and here are my takeaways:

1. There _was_ a change between python 3.1.4 and python 3.2.5 that caused the Merseinne Twister, python's default RNG, to return different results when given the same seed.
2. Since at least python 3.2.5, the python `random` module has returned consistent results
3. The Merseinne Twister (or at least python's implementation of it) appears to not behavior differently in a 32- or 64-bit build of python, even across operating systems.
4. More testing is needed to determine if different processor architectures (such as ARM) produce different results.

So all of that being said, I think this _particular_ use of `random` might be OK because the values themselves don't matter ... since we're asserting the sum of the source array against the sum of the warped array, all we need to know is that the pixel values are numerically close.  And even if we do have a situation where the results differ between runs, the seed is defined, so the only variables there are documented characteristics (python version, package versions, runner OS, etc.) of the GHA environment.

So I _think_ this might actually be OK!  But what do you think about that?

Reply:
Hey @phargogh , thanks for doing that analysis! That is super fun to scroll through. 

> since we're asserting the sum of the source array against the sum of the warped array, all we need to know is that the pixel values are numerically close.

Of course! I'm not quite sure why I thought we'd be using the random seeded inputs to make direct numerical comparisons against outputs.

>So I think this might actually be OK! But what do you think about that?

I think you're right and that this is fine. Thanks for looking into that though!